digdag-core/src/main/java/io/digdag/core/plugin/RemotePluginLoader.java

```diff
@@ -0,0 +1,190 @@
+
+    private static RepositorySystemSession newRepositorySystemSession(RepositorySystem system, Path localRepositoryPath)
+    {
+        DefaultRepositorySystemSession session = MavenRepositorySystemUtils.newSession();
+
+        LocalRepository localRepo = new LocalRepository(localRepositoryPath.toString());
+        session.setLocalRepositoryManager(system.newLocalRepositoryManager(session, localRepo));
+
+        //session.setTransferListener(new ConsoleTransferListener());
+        //session.setRepositoryListener(new ConsoleRepositoryListener());
+
+        return session;
+    }
+
+    private final RepositorySystem system;
+    private final RepositorySystemSession session;
+
+    public RemotePluginLoader(Path localRepositoryPath)
+    {
+        this.system = newRepositorySystem();
+        this.session = newRepositorySystemSession(system, localRepositoryPath);
+    }
+
+    @Override
+    public PluginSetFactory load(Spec spec)
+    {
+        if (spec.getDependencies().isEmpty()) {
+            return PluginSetFactory.empty();
+        }
+
+        ImmutableList.Builder<PluginFactory> builder = ImmutableList.builder();
+
+        List<RemoteRepository> repositories = getRepositories(spec);
+
+        for (String dep : spec.getDependencies()) {
+            // download artifacts, and/or resolve local-repository references to them
+            logger.debug("Loading plugin {}", dep);
+            List<ArtifactResult> artifactResults = resolveArtifacts(repositories, dep);
+
+            logger.debug("Classpath of plugin {}: {}", dep,
+                    artifactResults.stream().map(a -> a.getArtifact().getFile().toString())
+                    .collect(Collectors.joining(File.pathSeparator)));
+
+            ClassLoader pluginClassLoader = buildPluginClassLoader(artifactResults);
+            try {
+                ServiceLoader<PluginFactory> serviceLoader = ServiceLoader.load(PluginFactory.class, pluginClassLoader);
+                for (PluginFactory factory : serviceLoader) {
+                    builder.add(factory);
+                }
+            }
+            catch (ServiceConfigurationError ex) {
+                throw new RuntimeException("Failed to lookup io.digdag.spi.PluginFactory service from a dependency '" + dep + "'", ex);
+            }
+        }
+
+        return new PluginSetFactory(builder.build());
+    }
+
+    private ClassLoader buildPluginClassLoader(List<ArtifactResult> artifactResults)
+    {
+        ImmutableList.Builder<URL> urls = ImmutableList.builder();
+        for (ArtifactResult artifactResult : artifactResults) {
+            URL url;
+            try {
+                url = artifactResult.getArtifact().getFile().toPath().toUri().toURL();
+            }
+            catch (MalformedURLException ex) {
+                throw Throwables.propagate(ex);
+            }
+            urls.add(url);
+        }
+        return new PluginClassLoader(urls.build(), RemotePluginLoader.class.getClassLoader(),
```

Comment:
Sounds good.

I would prefer to be strict and throwing `NoClassDefFoundError` instead of loading non `io.digdag.spi` classes using parent/system classloader. I think it _is_ a big deal. The reason being that a plugin might otherwise unknowingly gain an indirect dependency on a class in the digdag parent/system classpath. That dependency might then break (best case with a `NoClassDefFoundError`, worst case with silently broken behavior) at some point in the future due to internal changes in digdag even though the plugin itself has not changed.